t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t

```diff
@@ -0,0 +1,597 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+
+use strict;
+use lib '.';
+use t::TestWasm::Lua;
+
+skip_no_openresty();
+
+plan tests => repeat_each() * (blocks() * 5);
+
+run_tests();
+
+__DATA__
+
+=== TEST 1: set_property_getter() - setting host properties getter fails on init_worker_by_lua
+--- wasm_modules: on_phases
+--- http_config
+    init_worker_by_lua_block {
+        local proxy_wasm = require "resty.wasmx.proxy_wasm"
+        local filters = {
+            { name = "on_phases" },
+        }
+
+        _G.c_plan = assert(proxy_wasm.new(filters))
+        assert(proxy_wasm.load(_G.c_plan))
+
+        local function getter(key)
+            if key == "my_key" then
+                return true, "my_value"
+            end
+            return false
+        end
+
+        assert(proxy_wasm.set_property_getter(getter))
+    }
+--- config
+    location /t {
+        rewrite_by_lua_block {
+            ngx.say("ok")
+        }
+    }
+--- response_body
+ok
+--- error_log eval
+[
+    qr/\[error\] .*? cannot set host properties getter outside of a request/,
+    qr/\[error\] .*? init_worker_by_lua(\(nginx\.conf:\d+\))?:\d+: could not set property getter/,
+]
+--- no_error_log
+[crit]
+
+
+
+=== TEST 2: set_property_getter() - setting host properties getter works on rewrite_by_lua
+--- wasm_modules: hostcalls
+--- http_config
+    init_worker_by_lua_block {
+        local proxy_wasm = require "resty.wasmx.proxy_wasm"
+        local filters = {
+            { name = "hostcalls" },
+        }
+
+        _G.c_plan = assert(proxy_wasm.new(filters))
+        assert(proxy_wasm.load(_G.c_plan))
+    }
+--- config
+    location /t {
+        proxy_wasm hostcalls 'on=response_body \
+                              test=/t/log/property \
+                              name=wasmx.my_property';
+
+        rewrite_by_lua_block {
+            local proxy_wasm = require "resty.wasmx.proxy_wasm"
+            assert(proxy_wasm.attach(_G.c_plan))
+
+            local function getter(key)
+                if key == "wasmx.my_property" then
+                    return true, "my_value"
+                end
+                return false
+            end
+
+            assert(proxy_wasm.set_property_getter(getter))
+
+            assert(proxy_wasm.start())
+            ngx.say("ok")
+        }
+    }
+--- response_body
+ok
+--- error_log eval
+[
+    qr/\[info\] .*? \[hostcalls\] on_response_body/,
+    qr/\[info\] .*? wasmx.my_property: my_value/,
```

Comment:
Recently I have stopped relying on `error_log` for my test cases because in my experience they result in false-positives, like this test atm. If not false-positive, they also fail to encompass larger edge-cases such as this one in my above comment. Furthermore, there is no way of knowing if `on_response_body` (the 1st regex) does happen before the 2nd one, or not. For rock-solid, tight test cases, I advise using `grep_error_log_out` at all times, extracting the necessary patterns such that even `wasmx.my_property: not found` string would also come up and make this test properly fail (at least in its current test).

Reply:
> For rock-solid, tight test cases, I advise using grep_error_log_out at all times, extracting the necessary patterns such that even wasmx.my_property: not found string would also come up and make this test properly fail (at least in its current test).

Yeah, I've followed some existing patterns when using `error_log` but I agree that `grep_error_log` is much better because of the checks for ordering. However, it can get pretty tricky/annoying to write a correct regexp for catching several lines in order and ensuring no unexpected repeats, etc. I'm thinking of adding a custom block handler where you would give a list of plain substrings (not regexes), and the test passes if the given substrings appear in that exact order with no additional occurrences of them. The subset of relevant lines could be filtered by a regex. It would make it lot easier to write solid tests.

EDIT: I left this comment open in the browser, went ahead, and implemented the block handler. PR incoming!

Reply:
Not opposed, but `grep_error_log` is still superior, because it may catch additional lines containing "the pattern" that may not be expected in the `grep_error_log_out` block... Plain substrings only solves the ordering aspect of `error_log`, but not this other issue which in my opinion, is well-worth the time spending crafting a matching regex and appropriate logs in the module & test filters.

Reply:
Actually I am not sure if I oppose or not. This would easily end up being used everywhere for convenience, but still not being as tight as `grep_error_log`...

Reply:
PR ready, do check it out — it's tighter than `grep_error_log`!